core/packages/contracts/src/AgentExecutor.sol

```diff
@@ -0,0 +1,32 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2023 Snowfork <hello@snowfork.com>
+pragma solidity 0.8.20;
+
+import {ParaID} from "./Types.sol";
+import {SubstrateTypes} from "./SubstrateTypes.sol";
+
+import {IERC20} from "./interfaces/IERC20.sol";
+import {SafeTokenTransfer, SafeNativeTransfer} from "./utils/SafeTransfer.sol";
+
+contract AgentExecutor {
+    using SafeTokenTransfer for IERC20;
+    using SafeNativeTransfer for address payable;
+
+    bytes32 internal constant COMMAND_TRANSFER_TOKEN = keccak256("transferToken");
+
+    function execute(address, bytes memory data) external {
+        (bytes32 command, bytes memory params) = abi.decode(data, (bytes32, bytes));
+        if (command == COMMAND_TRANSFER_TOKEN) {
+            (address token, address recipient, uint128 amount) = abi.decode(params, (address, address, uint128));
+            _transferToken(token, recipient, amount);
+        }
+    }
+
+    function transferNative(address payable recipient, uint256 amount) external {
+        recipient.safeNativeTransfer(amount);
+    }
+
+    function _transferToken(address token, address recipient, uint128 amount) internal {
+        IERC20(token).safeTransfer(recipient, amount);
```

Comment:
Will this be
`IERC20(token).safeTransferFrom(assetHubAgent, recipient, amount);` ?

Reply:
Nope.

The agent will invoke the `AgentExecutor.execute` function via `delegatecall`, and so `IERC20(token).safeTransfer` means that the tokens will be transferred from the agent to recipient.

See the `Gateway.agentExecute` function that implements this.

Reply:
Users definitely need to call `IERC20.approve` to allow the gateway contract to deposit the tokens into the agent.

See the test for sendNativeToken here: https://github.com/Snowfork/snowbridge/blob/7b7a037296a7698af28703cd525fd630ed282d8e/core/packages/contracts/test/Gateway.t.sol#L458